src/Rector/ClassMethod/AddGenericReturnTypeToRelationsRector.php

```diff
@@ -0,0 +1,184 @@
+    }
+
+    /**
+     * @param ClassMethod $node
+     */
+    public function refactor(Node $node): ?Node
+    {
+        if ($this->shouldSkipNode($node)) {
+            return null;
+        }
+
+        /** @var Node\Stmt[] $methodStatements */
+        $methodStatements = $node->stmts;
+
+        $methodReturnType = $node->getReturnType();
+
+        if ($methodReturnType === null) {
+            return null;
+        }
+
+        $methodReturnTypeName = $this->getName($methodReturnType);
+
+        if ($methodReturnTypeName === null) {
+            return null;
+        }
+
+        if (! $this->isObjectType(
+            $methodReturnType,
+            new ObjectType('Illuminate\Database\Eloquent\Relations\Relation')
+        )) {
+            return null;
+        }
+
+        $docInfo = $this->phpDocInfoFactory->createFromNode($node);
+        if (! $docInfo instanceof PhpDocInfo) {
+            return null;
+        }
+
+        // Return, if already has return type
+        if ($node->getDocComment() !== null && $docInfo->hasByName('return')) {
+            return null;
+        }
+
+        $returnStatement = $this->betterNodeFinder->findFirstInstanceOf($methodStatements, Return_::class);
+
+        if (! $returnStatement instanceof Return_) {
+            return null;
+        }
+
+        $relationMethodCall = $this->betterNodeFinder->findFirstInstanceOf($returnStatement, MethodCall::class);
+
+        if (! $relationMethodCall instanceof MethodCall) {
+            return null;
+        }
+
+        $relatedClass = $this->getRelatedModelClassFromMethodCall($relationMethodCall);
+
+        if ($relatedClass === null) {
+            return null;
+        }
+
+        $docInfo->addTagValueNode(
+            new ReturnTagValueNode(
+                new GenericTypeNode(
+                    new FullyQualifiedIdentifierTypeNode($methodReturnTypeName),
+                    [new FullyQualifiedIdentifierTypeNode($relatedClass)],
+                ),
+                ''
+            )
+        );
+
+        return $node;
+    }
+
+    private function getRelatedModelClassFromMethodCall(MethodCall $methodCall): ?string
+    {
+        if (count($methodCall->getArgs()) < 1) {
+            return null;
+        }
+
+        $argType = $this->getType($methodCall->getArgs()[0]->value);
+
+        if ($argType instanceof ConstantStringType) {
+            return $argType->getValue();
+        }
+
+        if (! $argType instanceof GenericClassStringType) {
+            return null;
+        }
+
+        $modelType = $argType->getGenericType();
+
+        if (! $modelType instanceof ObjectType) {
+            return null;
+        }
+
+        return $modelType->getClassName();
+    }
+
+    private function shouldSkipNode(ClassMethod $node): bool
+    {
+        if ($node->stmts === null) {
+            return true;
+        }
+
+        $classLike = $this->betterNodeFinder->findParentType($node, ClassLike::class);
```

Comment:
is it possible on `Interface_` and `Trait_` ?, If not, you can use specific node only if it only possible for 1 node, eg for `Class_` only:

```php
$class = $this->betterNodeFinder->findParentType($node, Node\Stmt\Class_::class);
```

Reply:
Yes, in can happen in trait also. I added a test case for that and update code to be more bulletproof.